### PR TITLE
setup.py uses cmake build system

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -188,28 +188,28 @@ target_include_directories(onnx PUBLIC ${ONNX_ROOT} "${CMAKE_CURRENT_BINARY_DIR}
 target_link_libraries(onnx PUBLIC onnx_proto)
 
 if(BUILD_PYTHON)
-  add_library(onnx_pybind MODULE "${ONNX_ROOT}/onnx/cpp2py_export.cc")
-  target_include_directories(onnx_pybind PRIVATE "${CMAKE_CURRENT_BINARY_DIR}" "${PROTOBUF_INCLUDE_DIRS}" "${PYTHON_INCLUDE_DIR}")
+  add_library(onnx_cpp2py_export SHARED "${ONNX_ROOT}/onnx/cpp2py_export.cc")
```
Comment (yinghai) [inline CMakeLists.txt]: `MODULE` is the recommended type as it's a Python plugin. https://cmake.org/cmake/help/v3.0/command/add_library.html

File: CMakeLists.txt [hunk 2]
```diff
@@ -188,28 +188,28 @@ target_include_directories(onnx PUBLIC ${ONNX_ROOT} "${CMAKE_CURRENT_BINARY_DIR}
...
   find_package(pybind11)
   if(pybind11_FOUND)
-    target_include_directories(onnx_pybind PRIVATE ${pybind11_INCLUDE_DIRS})
+    target_include_directories(onnx_cpp2py_export PRIVATE ${pybind11_INCLUDE_DIRS})
   else()
-    target_include_directories(onnx_pybind PRIVATE ${ONNX_ROOT}/third_party/pybind11/include)
+    target_include_directories(onnx_cpp2py_export PRIVATE ${ONNX_ROOT}/third_party/pybind11/include)
   endif()
-  target_link_libraries(onnx_pybind PRIVATE onnx_proto onnx)
+  target_link_libraries(onnx_cpp2py_export PRIVATE "-Wl,--whole-archive" onnx_proto onnx "-Wl,--no-whole-archive")
```
Comment (yinghai) [inline CMakeLists.txt]: Depending on the compiler we might need different flags. This is `GCC` specific.  Check
https://github.com/caffe2/caffe2/blob/53cc7dfe10fa1edd5d731654bea21d317375ffbf/cmake/public/utils.cmake#L10-L24

File: setup.py [hunk 3]
```diff
@@ -388,7 +480,7 @@ def pre_run(self):
 
 ext_modules = [
     create_extension(ONNXCpp2PyExtension,
-                     str('onnx.onnx_cpp2py_export'),
+                     str('libonnx_cpp2py_export'),
```
Comment (yinghai) [inline setup.py]: Can we have a name without `lib` just to indicate that it is a python module?

File: CMakeLists.txt (hunk 1 above)
Comment (anderspapitto) [reply CMakeLists.txt]: ++ meant to change it back

File: CMakeLists.txt (hunk 2 above)
Comment (anderspapitto) [reply CMakeLists.txt]: ++ will do

File: setup.py (hunk 3 above)
Comment (anderspapitto) [reply setup.py]: maybe. building with cmake via `add_library(foo)` forces it to be called 'libfoo' as far as I can tell - do you know how to disable that?

Comment (yinghai) [reply setup.py]: Try this? `set_target_properties(onnx_cpp2py_export PROPERTIES PREFIX "")`

**Comment by anderspapitto** [comment]
@onnxbot retest this please

File: CMakeLists.txt [hunk 4]
```diff
@@ -77,6 +85,10 @@ else()
 endif()
 add_definitions(${MY_ONNX_NAMESPACE})
 
+if(ONNX_ML)
+  add_definitions("-DONNX_ML=1")
```
Comment (yinghai) [inline CMakeLists.txt]: If we define `ONNX_ML`, when calling `gen_proto.py` we need `--ml` flag.

File: CMakeLists.txt [hunk 5]
```diff
@@ -189,28 +192,71 @@ target_include_directories(onnx PUBLIC ${ONNX_ROOT} "${CMAKE_CURRENT_BINARY_DIR}
...
+      from distutils import sysconfig
+      print(sysconfig.get_python_lib(prefix=''))
+  ")
+  # ---[ Options.
+  SET(PYTHON_LIB_REL_PATH "${PYTHON_SITE_PACKAGES}" CACHE STRING "Python installation path (relative to CMake installation prefix)")
+  message(STATUS "Using ${PYTHON_LIB_REL_PATH} as python relative installation path")
+  # Python extension suffix
+  # Try to get from python through sysconfig.get_env_var('EXT_SUFFIX') first,
+  # fallback to ".pyd" if windows and ".so" for all others.
+  pycmd(PY_EXT_SUFFIX "
```
Comment (yinghai) [inline CMakeLists.txt]: Since you hve `pycmd` now, you can use that on L92 too.

File: CMakeLists.txt [hunk 6]
```diff
@@ -189,28 +192,71 @@ target_include_directories(onnx PUBLIC ${ONNX_ROOT} "${CMAKE_CURRENT_BINARY_DIR}
...
+  elseif(MSVC)
+    # In MSVC, we will add whole archive in default.
+    target_link_libraries(onnx_cpp2py_export PRIVATE -WHOLEARCHIVE:$<TARGET_FILE:onnx>)
+    target_link_libraries(onnx_cpp2py_export PRIVATE ${PYTHON_LIBRARIES})
+  else()
+    # Assume everything else is like gcc
+    target_link_libraries(onnx_cpp2py_export PRIVATE "-Wl,--whole-archive" $<TARGET_FILE:onnx> "-Wl,--no-whole-archive")
+  endif()
+
+  target_link_libraries(onnx_cpp2py_export PRIVATE onnx)
```
Comment (yinghai) [inline CMakeLists.txt]: Why do we have this line again?

File: CMakeLists.txt [hunk 7]
```diff
@@ -242,13 +288,8 @@ if (WIN32)
         /wd4800 # disable warning type' : forcing value to bool 'true' or 'false' (performance warning)
         /wd4503 # identifier' : decorated name length exceeded, name was truncated
     )
-    if(${ONNX_USE_MSVC_STATIC_RUNTIME})
```
Comment (yinghai) [inline CMakeLists.txt]: Why do we remove this? We should leave this as optional and update the CI if necessary.

File: CMakeLists.txt [hunk 8]
```diff
@@ -189,28 +192,67 @@ target_include_directories(onnx PUBLIC ${ONNX_ROOT} "${CMAKE_CURRENT_BINARY_DIR}
 target_link_libraries(onnx PUBLIC onnx_proto)
 
 if(BUILD_PYTHON)
```
Comment (bddppq) [inline CMakeLists.txt]: I prefer to no support building python extension without going through setup.py. All these PY_EXT_SUFFIX, PYTHON_EXECUTABLE cmake variables should be directly passed from setup.py (passing via cmake flags).

File: setup.py [hunk 9]
```diff
@@ -28,13 +31,15 @@
 try:
     import ninja # noqa
     WITH_NINJA = True
+    WITH_NINJA = False
```
Comment (bddppq) [inline setup.py]: we can remove all these NINJA stuffs, basically delegate the ninja generation work to cmake and we only check whether `ninja` command is available.

File: setup.py [hunk 10]
```diff
@@ -291,17 +335,43 @@ def load(filename):
...
+            if os.name == 'nt':
+                debug_lib_dir = os.path.join(lib_path, "Debug")
+                release_lib_dir = os.path.join(lib_path, "Release")
+                if os.path.exists(debug_lib_dir):
+                    lib_path = debug_lib_dir
+                elif os.path.exists(release_lib_dir):
+                    lib_path = release_lib_dir
+            src = os.path.join(lib_path, filename)
+
+            if not os.path.exists(src):
```
Comment (bddppq) [inline setup.py]: We don't need this logic here, because the extensions list is fixed in onnx (aka. we are always building ['onnx_cpp2py_export']).

**Comment by yinghai** [comment]
Folks, once you make changes, please make new commit instead of squash and `git push -f` or it's difficult to track the progress of this PR. :)

File: appveyor.yml [hunk 11]
```diff
@@ -18,6 +18,14 @@ environment:
 
   matrix:
   # onnx-ml
+  - ONNX_ML: 1
+    ONNXPY_DIR: C:\Python35-x64
+    CONDA_PREFIX: C:\Miniconda35-x64
+
+  - ONNX_ML: 1
+    ONNXPY_DIR: C:\Python36-x64
```
Comment (bddppq) [inline appveyor.yml]: where is this used for?

File: appveyor.yml [hunk 12]
```diff
@@ -18,6 +18,14 @@ environment:
...
+    CONDA_PREFIX: C:\Miniconda36-x64
```
Comment (bddppq) [inline appveyor.yml]: this is not used anymore after this change.

File: appveyor.yml (hunk 11 above)
Comment (anderspapitto) [reply appveyor.yml]: looks like it's not, I'll try removing it

File: appveyor.yml (hunk 12 above)
Comment (anderspapitto) [reply appveyor.yml]: conda_prefix is added to PATH below

**Pull request merged**